src/linked_hash_map.rs

```diff
@@ -1677,6 +1749,173 @@ impl<'a, K, V> Drop for Drain<'a, K, V> {
+    #[inline]
+    fn muv(&mut self, at: impl FnOnce() -> NonNull<Node<K, V>>) {
+        self.cur = at().as_ptr();
+    }
+
+    /// Inserts the provided key and value before the current item. It checks if an entry
+    /// with the given key exists and, if so, replaces its value with the provided `1`
+    /// parameter from the given tuple. The key is not updated; this matters for types that
+    /// can be `==` without being identical.
+    ///
+    /// If the entry doesn't exist, it creates a new one. If a value has been updated, the
+    /// function returns the *old* value wrapped with `Some`  and `None` otherwise.
+    #[inline]
+    pub fn insert_before(&mut self, tuple: (K, V)) -> Option<V>
+    where
+        K: Eq + Hash,
+        S: BuildHasher,
+    {
+        let before = unsafe { NonNull::new_unchecked(self.cur) };
+        self.insert(tuple, || before)
+    }
+
+    /// Inserts the provided key and value after the current item. It checks if an entry
+    /// with the given key exists and, if so, replaces its value with the provided `1`
+    /// parameter from the given tuple. The key is not updated; this matters for types that
+    /// can be `==` without being identical.
+    ///
+    /// If the entry doesn't exist, it creates a new one. If a value has been updated, the
+    /// function returns the *old* value wrapped with `Some`  and `None` otherwise.
+    #[inline]
+    pub fn insert_after(&mut self, tuple: (K, V)) -> Option<V>
+    where
+        K: Eq + Hash,
+        S: BuildHasher,
+    {
+        let before = unsafe { (*self.cur).links.value.next };
+        self.insert(tuple, || before)
+    }
+
+    // Inserts `item` immediately before the element returned by the `before` closure function.
+    #[inline]
+    fn insert(&mut self, tuple: (K, V), before: impl FnOnce() -> NonNull<Node<K, V>>) -> Option<V>
+    where
+        K: Eq + Hash,
+        S: BuildHasher,
+    {
+        let (key, value) = tuple;
+        unsafe {
+            let hash = hash_key(self.hash_builder, &key);
+            let i_entry = self
+                .table
+                .find_entry(hash, |o| (*o).as_ref().key_ref().eq(&key));
+
+            match i_entry {
+                Ok(occupied) => {
+                    let mut node = *occupied.into_mut();
+                    let pv = mem::replace(&mut node.as_mut().entry_mut().1, value);
+                    detach_node(node);
+                    attach_before(node, before());
+                    Some(pv)
+                }
+                Err(_) => {
+                    let mut new_node = allocate_node(self.free);
+                    new_node.as_mut().put_entry((key, value));
+                    let hash_builder = self.hash_builder;
+                    self.table.insert_unique(hash, new_node, move |k| {
+                        hash_key(hash_builder, (*k).as_ref().key_ref())
+                    });
+                    attach_before(new_node, before());
```

Comment:
This one's *really* pedantic, but I *think* everywhere where an insert into the table is actually performed, the node is attached to the linked list *before* inserting into the table.

The reason this matters is for behavior under panics, it saves a memory leak if the hash function panics (the linked list is considered the proper *owner* of the values rather than the table).